### DISCUSSION TITLE
LastMotionTime fails to update

### DISCUSSION THREAD

**Discussion by snigehere** [tsightler/ring-mqtt#448] [Support Q&A]
I am having issues with the add on where the last motion timestamp is not being updated.  In the log I can see other activity going on such as snapshot and info updates.  If I stop and restart the ring mqtt addon it corrects the issue for a while but stops updating last motion time again after a period of time (everything else continues to work OK).

Restarting the mosquitto broker does not make any difference.  

Attached is a picture showing info from a number of places - the ring app showing a selection of motion detections, the sensor in HA, the sensor history showing no activity for many hours and the ring mqtt log showing transactions.  

Is this an issue on the ring end or is there some issue with the add on failing to collect the data?  

![ring mqtt last motion issue](https://user-images.githubusercontent.com/104912027/187538954-7a1b0627-f3bf-40d7-966c-7cca2e61329c.jpg)

I previously tried to export the log (for a different issue) using the instructions for writing to a file - the file was written, and showed it in .\ (confirmed with LS command) but I was not able to find it through studio code editor or file editor.  My skills in linux/docker are zero - I am guessing its a file authority issue but really have no idea.  If logs are needed please assist with "docker for dummies" guidance.

I run on windows 10 using virtualbox - I have the file explorer and studio code server addons together with samba share (with the backup and config folders mapped to windows) 

Thanks for any help

**Comment by tsightler** [comment]
Are you using any Ring features, such as Smart Alerts, which suppress notifications?  Ring-mqtt depends on push notifications so it will only know about events if there are notifications, so there can absolutely be events in the log that are not in ring-mqtt.

It's not really clear to me what you are asking about, you are talking about the timestamp, but realistically, are you simply not receive motion notifications at all?

Basically, without full logs, it's pretty much impossible to troubleshoot something like this.  I'm not sure what docker help I can provide, running "docker logs" command is pretty straightforward.  Are you able to run the "docker log" command from the ssh session without the redirect to a file and see the logs?

**Reply by snigehere** [reply]
Hi - Thanks for the prompt response - No - alerts are not suppressed - I get the notification on my phone so know motion is detected.

I use binary_sensor.front_door_doorbell_motion_2 (which is the motion sensor for my front door bell) to trigger a push of the camera feed to my display panel so we can see who is at the door.  
I show the last motion timestamp on my camera dashboard so we can see when each camera last detected motion.

... so I am not getting a motion trigger (therefor no push of the live feed to the panel, and the timestamp is not updated. 

I followed the steps here https://github.com/tsightler/ring-mqtt-ha-addon/blob/main/GET-LOGS.md to get the logs

I issued this command  
docker logs addon_03cabcc9_ring_mqtt 2>&1 | gzip > ring-mqtt.log.gz
and a file was produced - I can see if when I list the files but I have no idea how to find it and get it out so I can send it anywhere.  

The guidance "Using a tool like WinSCP, an addon like File Editor, or any other method you prefer, to copy/download the ring-mqtt.log.gz file to your computer" unfortunately did not help me - I installed the file editor add on but could not find the created log file.

I can see it when I use find on the ssh terminal

![terminal](https://user-images.githubusercontent.com/104912027/187554583-71988871-9a35-46da-a275-0abe6cb5ecf4.jpg)

but not via file editor or studio .... I have no clue why - as I said I assume its some sort of read authority issue.....

**Reply by tsightler** [reply]
You said above you have the Samba share setup with the /backup and /config folders, so you should be able to just copy/move the file to one of these two folders and pull it from the share.  For example something like:

`mv ring-mqtt.log.gz config`

Or you could just change directory to the given folder and then dump the logs.

**Comment by snigehere** [comment]
Great - thanks - that was exactly what I needed... 

I now have the logs that cover a failure - I would prefer not to post publicly as its quite big and contains a lot of detail but happy to share directly.  I have attached a spreadsheet that includes the camera entries in the time period so a bit more detail than below but still just limited to the doorbell camera (nb I have changed the id/password detail to xxx@xxx)

So today @ 08:44 (local, 07:44 UTC) there was a motion event at the door - I can see that being picked up in the log - I removed the info from all the other devices to make it easier to post.  In the first event we see "new motion event" and the motion notification ... in the second we see "new motion event" but no motion push:

2022-08-31T07:43:48.879Z ring-mqtt  [32m[Front Door Doorbell]  [39mNew motion event detected, updating the event recording URL
2022-08-31T07:43:49.178Z ring-attr  [32m[Front Door Doorbell]  [39m [34mring/79a9a311-601a-4df6-a696-8d804cfbca75/camera/083a882613b0/event_select/attributes  [39m [36m<recording_url_masked> [39m

<snip>

2022-08-31T07:44:02.477Z ring-mqtt  [32m[Front Door Doorbell]  [39mReceived motion push notification, expires in 180 seconds
2022-08-31T07:44:02.478Z ring-mqtt  [32m[Front Door Doorbell]  [39mRequesting motion snapshot using notification image UUID: bc5a8ce7-8797-4d75-b209-09e1255de63e:49454583
2022-08-31T07:44:02.479Z ring-mqtt  [32m[Front Door Doorbell]  [39m [34mring/79a9a311-601a-4df6-a696-8d804cfbca75/camera/083a882613b0/motion/state  [39m [36mON [39m
2022-08-31T07:44:02.479Z ring-attr  [32m[Front Door Doorbell]  [39m [34mring/79a9a311-601a-4df6-a696-8d804cfbca75/camera/083a882613b0/motion/attributes  [39m [36m{"lastMotion":1661931842,"lastMotionTime":"2022-08-31T07:44:02Z","personDetected":false,"motionDetectionEnabled":true} [39m

The next motion event at the front door was 20 minutes later

2022-08-31T08:04:15.516Z ring-mqtt  [32m[Front Door Doorbell]  [39mNew motion event detected, updating the event recording URL
2022-08-31T08:04:15.813Z ring-attr  [32m[Front Door Doorbell]  [39m [34mring/79a9a311-601a-4df6-a696-8d804cfbca75/camera/083a882613b0/event_select/attributes  [39m [36m<recording_url_masked> [39m

<snip>

2022-08-31T08:04:26.917Z ring-mqtt  [32m[Front Door Doorbell]  [39mRequesting an updated interval snapshot

[ring_mqtt_issue_31-08-2022.xlsx](https://github.com/tsightler/ring-mqtt/files/9461373/ring_mqtt_issue_31-08-2022.xlsx)

**Reply by tsightler** [reply]
Thanks, please sends logs to my email if you don't mind, same as my github username at gmail.

**Reply by tsightler** [reply]
I'm not sure if you sent them but I didn't see any logs.  Regardless, if you get a chance, could you please try the dev branch see if this helps with the issue.  I've updated it to use custom version of push-receiver that has some updates that will hopefully help with the issue.

**Comment by snigehere** [comment]
Thanks - sorry the logs didn't arrive - I see them in my sent log Wed, 31 Aug, 15:28 (17 hours ago) - I have sent them again - gmail shows your picture when I type in the ID so assume that is correct. 

Thanks for the clear documentation and very easy process to install the dev branch - I changed the configuration "branch" item to "dev" and responded to the restart message - the log shows its loaded 5.0.4 beta 1

I will update you on my results

Log messages:
Updating ring-mqtt to the dev version...
Cloning into 'ring-mqtt-dev'...
Installing node module dependencies, please wait...
The ring-mqtt-dev has been updated.
cont-init: info: /etc/cont-init.d/ring-mqtt.sh exited 0
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
services-up: info: copying legacy longrun ring-mqtt (no readiness notification)
s6-rc: info: service legacy-services successfully started

ring-mqtt.js version: 5.0.4-beta.1
Node version v16.16.0
NPM version 8.10.0
git version 2.36.2

Running ring-mqtt...

**Reply by tsightler** [reply]
Thanks, I received them this time around.  Unfortunately there is very little to go on as the logs look perfect.  Let's see if the behavior with the updated push-receiver package is better.  If you still see the behavior I may need to add some additional logging.  The problem is, this issue appears to be in these underlying components, most likely push-receiver, which I don't maintain, so this is going to be a tough one overall, if the updated push-receiver doesn't fix it.

**Comment by snigehere** [comment]
Hi - thanks for the update - unfortunately it has not solved the issue - I have multiple movements detected on two cameras where there is no push notification and no motion shown in the history in HA ... 

These two pictures show the HA sensor history side by side with the ring history 

![front_door_2022-09-02](https://user-images.githubusercontent.com/104912027/188038674-28e1e329-4baa-453e-b377-9f2d6ccc144d.jpg)
![back_garden_2022-09-02](https://user-images.githubusercontent.com/104912027/188038677-dd17e932-0e1d-4ebe-9188-bc43fb2419fb.jpg)

The front door camera "caught up" .. eg earlier today it was not picking up the movement but at 20:20 local the last motion updated to be in line with ring ... I have seen this before - it seems to miss notifications for a while then wake up again and pick one up.  You will see that the back garden notification is still not updated.  The "unavailable" you can see on the back garden camera is when I installed the dev version earlier today. 

In the past I have used both the ring cloud integration and the mqtt addon ... I have only recently got around to starting to use the functions like using the motion trigger to push the video to a tablet when motion is detected.  I began to see missing triggers so removed the ring cloud integration in case there was some issue trying to run both (that's also why the sensors have a "_2" value shown).

**Reply by tsightler** [reply]
Unfortunately, I don't know what else to do about it.  There are absolutely zero errors in the logs, no indication of anything being wrong, and I can't reproduce it with the dozens of cameras my account has access to.  The ring-mqtt code itself is does not handle the communication with Ring, so there's little left for me to do, especially since I can't reproduce this behavior.  I guess I'll just have to say that ring-mqtt won't work for you.

**Reply by snigehere** [reply]
hmm ... thanks for the feedback - can you give some more detail on the data flow please - you say that your code does not handle the coms - what does?  There is nothing strange or odd about my setup - I have my energy data flowing via mqtt and that does not give any issues..  My setup is a virtual box running on windows 10 like many other people.  My ring setup is straight forward, I have fibre broadband with a fast speed connection.  all the components of HA are up to date.

Is it just me or do others have "odd issues" that cause problems?

**Reply by tsightler** [reply]
I don't know if others have the issue, but so far you're the only one that has reported it since 5.x was released and switched to the push method for notifications.  I estimate this project has ~1500 active users at this point (perhaps more, that's pretty conservative based on docker pull counts and other metrics), and typically, if there is a common issue, I'd get multiple reports a week about it so, based on that, I assume the issue is relatively isolated.  I mean, I'm also a user of this project, I wrote it for myself, not for other people, so if it was constantly missing stuff, I'd be annoyed by it too, but I run 2 instances 24x7 and I have basically never seen this problem (I may have seen it once, but it was after a network outage so somewhat explainable).

That being said, I'm not saying that there's not a problem, I'm just saying that the problem likely lives below the ring-mqtt code itself, most likely in ring-client-api.  The ring-mqtt source doesn't implement the code that talks to the Ring API, that task is handled by ring-client-api, which is another project maintained by the author of the ring-homebridge plugin, and that project has even more users than ring-mqtt.  Not only that, but ring-client-api is also the basis for probably a dozen other projects, including another major video streaming project, Skrypted.  If a problem is widespread, then the issues usually pop up on ring-client-api/ring-homebridge first, and there is indeed an issue open there for similar behavior, but most of the people that have chimed in there are actually having an issue with the way push notifications work vs the previous polling method, and it's still only a handful of comments over the weeks that it's been open, so it doesn't seem widespread.

The use of push-notifications was only implemented in ring-client-api about 6 months ago, and, while it has some huge benefits, especially for battery cameras, it's not completely surprising that it would go through some growing pains.  Unfortunately, the maintainer of ring-client-api decided to rip and replace the old polling method with the push notifications, which is certainly his prerogative, but that means the only options are for me to maintain my own fork of ring-client-api, which I just don't have the time or interest in doing, or to wait for issues to be fixed there.

I had hoped that the forked version of push-receiver (yet another project that ring-client-api uses to register for the push notifications) might fix the problem, but it seems it didn't.  However, I did go glance at the code and it's possible I need to modify it slightly to enable the heartbeat function as I previously assumed it was enabled by default.  I was hoping that the heartbeat which was implemented in this forked version might fix the issue since, without it, it can take Linux an incredibly long time to recognize when a TCP connection is closed (up to two hours).

This somewhat matched the behavior you report.  My theory was that the connection is closed on the remote side for whatever reason but Linux isn't aware of it so it doesn't know it needs to close/destroy that socket and reconnect.  During this period no notifications are received, but finally, after 7200 seconds (2 hours) Linux decides to send a TCP heartbeat, realizes the socket is closed, re-establishes the connection and then notifications "catch up".  It's just a theory, but it seems to match.

Why it might be different for you vs me vs others, well, who knows, you may have a very normal local setup but there could be something in your firewall, or your ISP (like CGNAT) that is impacting long-lived TCP connections, which would largely go unnoticed for other use cases.  I mean, I really don't know but, unfortunately, I just don't have time to dig into every users specific case as this is just a personal project that I wrote for myself and threw on the internet because I thought others might find it interesting, I never expected it to have 1000's of users that want me to help them resolve problems with it.  I mean, I do try, but sometimes, especially when the problems are in dependent libraries rather than the code I wrote, I just have to recognize that there's not much I can do about it.

That being said, I do need to investigate this heartbeat thing a little more, as I had assumed this forked version enabled the heartbeat by default, but looking at it a little more deeply, it seems like it perhaps needs to be explicitly enabled, so maybe I can push one more development version and see if that helps.  If not, I really will be at the end of my options.

**Reply by tsightler** [reply]
If you want you can give the latest dev another spin, I've made 100% sure that heartbeat is enabled in the code.  If that still doesn't work I do have one more possible idea to try with a minor tweak to push-receiver.  Right now push-receiver does use a socket and calls setKeepalive, however, it doesn't specify an interval so it uses the system default, which is that crazy two hours, so maybe setting that lower might help.

BTW, you don't have to just believe me about the entire issues with TCP connection timeouts with routers and carriers and such, there are plenty other sources about problems with GCM notifications and various routers and such, the problem happens even with Andriod devices and there's an app to tun the GCM heartbeat called Heartbeat fixer.  Just one detailed thread about it:

https://groups.google.com/g/android-gcm/c/Y33c9ib54jY/m/YxnCkaPRHRQJ

So this is why I've enabled the heartbeat in this latest dev push, maybe that will fix it, but it's still just kind of a guess.  I suppose I could modify the ring-client-api to enable logging in the push receiver and see if we can find out what is going on in more detail.

**Comment by snigehere** [comment]
Hi - Many thanks for the very detailed explanation - it helps me understand a bit more about what's involved.  
I have restarted and beta2 is now installed 
ring-mqtt.js version: 5.0.4-beta.2
Node version v16.16.0
NPM version 8.10.0
git version 2.36.2
I do appreciate that much of hat is available in HA is peoples personal projects to solve their needs and that these are then shared and I understand that everyone is giving their time generously to help solve other peoples issues.  
I am happy to take any guess and try it.  The addon is great and so I will continue to use it and hope at some point something will change!  I worked as a solution architect for the past 25 year and know that tracking down isolated corner cases is a real challenge when there is a big team involved.

**Reply by tsightler** [reply]
Thanks for being willing to test and it's not that I'm not interested in digging into even the lower level libraries, it's just that I don't have the time to do it much these days.  Unfortunately, due to that lack of time, I've had to focus only on addressing issues that are within the codebase that I directly control.  It's also been worse due to the fact that so many people open issues that simply aren't even code related, and I definitely don't have time to help people figure out why their Ring camera is too slow starting a stream (reproduced in the Ring app), or why their system keeps losing network.

That being said, when people are willing to dig in and try to figure it out, then I'll usually jump in with both feet to try to help, even if the issue is in underlying libraries that I don't maintain.  I'll be interested to hear how it goes with the current beta code.

**Reply by snigehere** [reply]
thanks - I will get back early next week as I am away on business for a few days
